File: src/collections_and_containers/cpp/stack/src/stack.cpp
#include "stack.h"
#include <stdexcept>

static constexpr int DEFAULT_INIT_SIZE = 16;

template <class T>
Stack<T>::Stack()
    : data(new T[DEFAULT_INIT_SIZE]), n(0), maxSize(DEFAULT_INIT_SIZE) {}

template <class T> Stack<T>::Stack(const Stack &other) : data(nullptr) {
  copy(other);
}

template <class T> Stack<T>::~Stack() { delete[] data; }

template <class T> void Stack<T>::copy(const Stack &other) {
  n = other.n;
  maxSize = other.maxSize;
  data = new T[maxSize];

  for (int i = 0; i < n; i++) {
    data[i] = other.data[i];
  }
}

template <class T> void Stack<T>::expand() {
  int newMaxSize = maxSize * 2;
  T *newData = new T[newMaxSize];

  for (int i = 0; i < n; i++) {
    newData[i] = data[i];
  }

  delete[] data;
  data = newData;
  maxSize = newMaxSize;
}

template <class T> T Stack<T>::pop() {
  if (n <= 0)
    throw std::out_of_range("Stack is empty");

  return data[--n];
}

template <class T> void Stack<T>::push(T element) {
  if (n >= maxSize)
    expand();

  data[n++] = element;
}

template <class T> T Stack<T>::top() const {
  if (n <= 0)
    throw std::out_of_range("Stack is empty");

  return data[n - 1];
}

template <class T> int Stack<T>::size() const { return n; }

template <class T> bool Stack<T>::isEmpty() const { return n == 0; }

template <class T> void Stack<T>::clear() { n = 0; }

template <class T> Stack<T> &Stack<T>::operator=(const Stack<T> &other) {
  if (this != &other) {
    delete[] data;
    copy(other);
  }
  return *this;
}

template class Stack<int>;
template class Stack<float>;
template class Stack<double>;
template class Stack<char>;

File: src/collections_and_containers/python/stack/src/stack.py
"""Stack implementation using a list."""

from __future__ import annotations

from typing import Any


class Stack:
    """A LIFO (Last In, First Out) stack data structure."""

    def __init__(self, max_size: int = 1000) -> None:
        self.stack: list[Any] = []
        self.max_size = max_size

    def empty(self) -> bool:
        """Check if the stack is empty."""
        return len(self.stack) == 0

    def __eq__(self, other: object) -> bool:
        if isinstance(other, Stack):
            return self.stack == other.stack
        return False

    def pop(self) -> Any:
        """Remove and return the top element."""
        if self.empty():
            raise IndexError("Cannot pop from an empty stack")
        data = self.stack[-1]
        del self.stack[-1]
        return data

    def push(self, data: Any) -> None:
        """Add an element to the top of the stack."""
        if len(self.stack) == self.max_size:
            raise IndexError("Stack is full")
        self.stack.append(data)

    def top(self) -> Any:
        """Return the top element without removing it."""
        if self.empty():
            raise IndexError("Cannot peek at an empty stack")
        return self.stack[-1]

    def size(self) -> int:
        """Return the number of elements in the stack."""
        return len(self.stack)
